exercises/10-exercise.py

```diff
@@ -0,0 +1,97 @@
+'''
+Author: Jorge Diaz Jr
+
+Exercise 10
+-----------
+Get file size
+
+'''
+import numpy as np
+import os
+from functools import wraps
+
+# Some bluesky modules
+from bluesky import RunEngine
+from bluesky.plans import count
+
+# simulated devices from ophyd
+from ophyd.sim import (img, det1, det2, NumpySeqHandler, motor1,
+                       SynSignalWithRegistry)
+
+# databroker modules
+from databroker.tests.utils import temp_config
+from databroker import Broker
+
+
+class NumpySeqHandler:
+    specs = {'NPY_SEQ'}
+
+    def __init__(self, filename, root=''):
+        self._name = os.path.join(root, filename)
+
+
+    def __call__(self, index):
+        return np.load('{}_{}.npy'.format(self._name, index))
+
+
+    def get_file_list(self, datum_kwarg_gen):
+        #This method is optional. It is not needed for access, but for export.
+        return ['{name}_{index}.npy'.format(name=self._name, **kwargs['index']) #**kwargs
+                for kwargs in datum_kwarg_gen]
+
+
+    def get_file_size(self, datum_kwarg_gen):
+        resource = db.reg.resource_given_datum_id(datum_kwarg_gen['datum_id'])
+        fpath = resource['root'] + "/" + resource['resource_path'] + \
+            '_' + str(datum_kwarg_gen['datum_kwargs']['index']) + '.npy'
+        size = os.path.getsize(fpath)
+        
+        return size
+
+
+events = []
+datum_ids = []
+resources = []
+datums = []
+
+db = Broker.from_config(temp_config())
+
+reg_handler = db.reg.register_handler("NPY_SEQ", NumpySeqHandler)
+
+# img is a simulated detector. It needs to know where reg is located
+RE = RunEngine({})
+
+# subscribe writing to database to the RunEngine
+RE.subscribe(db.insert)
+
+# counting images
+for i in range(10):
+    uid, = RE(count([img, det1, det2]))
+
+#extracting the events from each header doc in the files generated
+for i in range(10):
+    hdr = db[-i - 1]
+    event = hdr.events()
+    events.append(next(event))
+
+#extracting the datum ids from each event that was generated
+for i in range(len(events)):
+    datum_ids.append(events[i]['data']['img'])
+
+# call these functionos to get the resource and datum
```

Comment:
we should create a set of resources instead and ignore the datums.
```
resources = set()
```

and further down 
```
resources.add(resource) 
```
This should work. If there are issues, then it is because resource is not hashable.
Basically, elements in a set need to be hashable (so we can hash them and find out if they're unique to other elements in a set). Things that are hashable are usually immutable types (objects that can't be modified), such as tuples, numbers. Lists, on the other hand are not hashable (since we can add to lists)

If you can't make a set of resources, then just make a set of resource id's. Making the set will just make things easier in terms of finding unique resources.

Reply:
In order for me to create an object for the class, I need the 'resource path' and 'root' from resource document, since a resource is not hashable, I did what you said, to only extract the resource id's, but now when I'm creating the file-handler object, I need both the resource path and the root. Can I convert the set to a list and work with that or that doesn't matter, or should I just save the resources in order to access the resource path and root.